File: Algo/Leetcode/054SpiralMatrix.cpp
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        vector<int> res;
        int m = matrix.size();
        if (m==0) return res;
        int n = matrix[0].size();
        if (n==0) return res;

        int start=0;
        while(start*2 < m && start*2 < n){
            printMatrixByStart(start, m, n, matrix, res);
            start++;
        }
        return res;
    }

    void printMatrixByStart(int start, int m, int n, vector<vector<int> > matrix, vector<int>& res){
        int end_row = m - start - 1;
        int end_col = n - start - 1;
        for(int i=start; i<=end_col; i++)
            res.push_back(matrix[start][i]);

        for(int i=start+1; i<=end_row; i++)
            res.push_back(matrix[i][end_col]);

        if (end_row > start){
            for(int i=end_col-1; i>=start; i--)
                res.push_back(matrix[end_row][i]);
        }

        if(end_col > start){
            for(int i=end_row-1; i>start; i--)
            res.push_back(matrix[i][start]);
        }
    }
};

File: Algo/Leetcode/054SpiralMatrix.py
class Solution(object):
    def spiralOrder(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: List[int]
        """
        n = len(matrix)
        m = len(matrix[0])
        l, r, u, d = 0, m-1, 0, n-1
        res = []
        while l <= r and u <= d:
            if l == r:
                res.extend(printHorizontal(matrix, u, l, r))
                if u < d:
                    res.extend(printVetical(matrix, l, u+1, d))
                return res
            else:
                res.extend(printHorizontal(matrix, u, l, r))
                if u < d:
                    res.extend(printVetical(matrix, r, u+1, d))
                    res.extend(printHorizontal(matrix, d, r-1, l))
                    if u < d-1:
                        res.extend(printVetical(matrix, l, d-1, u+1))
                    l, r, u, d = l+1, r-1, u+1, d-1
                else:
                    return res

        return res

def printHorizontal(mat, i, j1, j2):
    if j2 >= j1:
        return [mat[i][k] for k in range(j1, j2+1)]
    else:
        return [mat[i][k] for k in range(j1, j2-1, -1)]

def printVetical(mat, j, i1, i2):
    if i2 >= i1:
        return [mat[k][j] for k in range(i1, i2+1)]
    else:
        return [mat[k][j] for k in range(i1, i2-1, -1)]